src/assets/scripts/client/aircraft/AircraftModel.js

```diff
@@ -1457,18 +1545,48 @@ export default class AircraftModel {
             //     return;
 
             case MCP_MODE.ALTITUDE.VNAV: {
-                const waypointAltitude = this.fms.currentWaypoint.altitudeRestriction;
-                const waypointHasAltitude = waypointAltitude !== -1;
+                const hardRestrictedWaypointModel = this.fms.nextHardAltitudeRestrictedWaypoint;
 
-                if (!waypointHasAltitude) {
-                    return this.mcp.altitude;
+                if (_isNil(hardRestrictedWaypointModel)) {
+                    return;
                 }
 
+                let waypointAltitude = hardRestrictedWaypointModel.altitudeMaximum;
+
                 if (this.flightPhase === FLIGHT_PHASE.TAKEOFF || this.flightPhase === FLIGHT_PHASE.CLIMB) {
```

Comment:
mmm, so much cleaner.  And once we get to writing tests for this (sooner rather than later I hope), you will thank me.  This is so much easier to test now.